src/autopas/selectors/tuningStrategy/PredictiveTuning.h

```diff
@@ -0,0 +1,370 @@
+    // get all traversals of the container and restrict them to the allowed ones
+    const std::set<TraversalOption> &allContainerTraversals =
+        compatibleTraversals::allCompatibleTraversals(containerOption);
+    std::set<TraversalOption> allowedAndApplicable;
+    std::set_intersection(allowedTraversalOptions.begin(), allowedTraversalOptions.end(),
+                          allContainerTraversals.begin(), allContainerTraversals.end(),
+                          std::inserter(allowedAndApplicable, allowedAndApplicable.begin()));
+
+    for (auto &cellSizeFactor : allowedCellSizeFactors)
+      for (auto &traversalOption : allowedAndApplicable) {
+        for (auto &dataLayoutOption : allowedDataLayoutOptions) {
+          for (auto &newton3Option : allowedNewton3Options) {
+            _searchSpace.emplace(containerOption, cellSizeFactor, traversalOption, dataLayoutOption, newton3Option);
+          }
+        }
+      }
+  }
+
+  AutoPasLog(debug, "Points in search space: {}", _searchSpace.size());
+
+  if (_searchSpace.empty()) {
+    autopas::utils::ExceptionHandler::exception("PredictiveTuning: No valid configurations could be created.");
+  }
+
+  for (auto &configuration : _searchSpace) {
+    std::vector<std::pair<int, size_t>> vector;
+    _traversalTimesStorage.emplace(configuration, vector);
+  }
+
+  _currentConfig = _searchSpace.begin();
+}
+
+void PredictiveTuning::selectPredictedConfigurations() {
+  if (_searchSpace.size() == 1 || _timer == 0 || _timer == 1) {
+    _optimalSearchSpace = _searchSpace;
+    return;
+  }
+
+  predictConfigurations();
+
+  auto optimum = std::min_element(_traversalPredictions.begin(), _traversalPredictions.end(),
+                                  [](std::pair<Configuration, size_t> a, std::pair<Configuration, size_t> b) -> bool {
+                                    return a.second < b.second;
+                                  });
+
+  _optimalSearchSpace.emplace(optimum->first);
+
+  // selects configurations that are near the optimal prediction or have not been tested in the last five iterations
+  for (auto &configuration : _searchSpace) {
+    auto vector = _traversalTimesStorage[configuration];
+    if ((_timer - vector[vector.size() - 1].first) >= 5 ||
+        ((float)_traversalPredictions[configuration] / optimum->second <= 1.2)) {
+      _optimalSearchSpace.emplace(configuration);
+    }
+  }
+}
+
+void PredictiveTuning::predictConfigurations() {
+  linePrediction();
+  /*switch (method) {
+  * case line: linePrediction(); break;
+  * default:
+
+  }*/
+}
+
+void PredictiveTuning::linePrediction() {
+  for (auto &configuration : _searchSpace) {
+    auto vector = _traversalTimesStorage[configuration];
+    auto tmp1 = vector[vector.size() - 1];
+    auto tmp2 = vector[vector.size() - 2];
+
+    auto prediction = tmp1.second + (tmp1.second - tmp2.second) / (tmp1.first - tmp2.first) * (_timer - tmp1.first);
+    _traversalPredictions[configuration] = prediction;
```

Comment:
```suggestion
    _traversalPredictions[configuration] = tmp1.second + (tmp1.second - tmp2.second) / (tmp1.first - tmp2.first) * (_timer - tmp1.first);
```
No need for a temporary here.
Please add a small comment on what this calculation is. E.g. like `t1 + (t1-t2) / (iter1 - iter2) * (tuningPhase - iter1)` or something more understandable....

Reply:
So if I understand this correctly, you make your prediction base on the number of the tuning Iteration (`_timer`). However, wouldn't it be more precise if it was based on the actual iteration number, since the tuning iterations might not be equally long?